app/controllers/actions/PrePopDataAction.scala

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.actions
+
+import services.PsrRetrievalService
+import play.api.mvc.ActionTransformer
+import config.Constants.PREPOP_PREFIX
+import models.UserAnswers
+import repositories.SessionRepository
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
+import models.requests.{AllowedAccessRequest, DataRequest}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+import javax.inject.{Inject, Singleton}
+
+@Singleton
+class PrePopDataAction @Inject()(sessionRepository: SessionRepository, psrRetrievalService: PsrRetrievalService)(
+  implicit val ec: ExecutionContext
+) {
+
+  def apply(optLastSubmittedPsrFbInPreviousYears: Option[String]): ActionTransformer[DataRequest, DataRequest] =
+    new ActionTransformer[DataRequest, DataRequest] {
+
+      override protected def transform[A](existingDataRequest: DataRequest[A]): Future[DataRequest[A]] =
+        optLastSubmittedPsrFbInPreviousYears
+          .map { lastSubmittedPsrFbInPreviousYears =>
+            implicit val hc: HeaderCarrier =
+              HeaderCarrierConverter.fromRequestAndSession(existingDataRequest, existingDataRequest.session)
+            val allowedAccessRequest = existingDataRequest.request
+            for {
+              prePopReturn <- psrRetrievalService.getAndTransformStandardPsrDetails(
+                optFbNumber = Some(lastSubmittedPsrFbInPreviousYears),
+                fallBackCall = controllers.routes.OverviewController.onPageLoad(existingDataRequest.srn)
+              )(
+                hc = implicitly,
+                ec = implicitly,
+                request = DataRequest[A](allowedAccessRequest, emptyUserAnswers(allowedAccessRequest))
+              )
+              cleanedPrePopUa = cleanUpPrePop(prePopReturn)
+              _ <- sessionRepository.set(cleanedPrePopUa.copy(id = PREPOP_PREFIX + prePopReturn.id))
```

Comment:
Should the prePopulated UA get deleted somewhere. Now if you navigate back to the overview page on the UI and go down on the View and change route, request.prePopUserAnswers is still there.

Reply:
If I consider this question and above one together, we may copy and paste over current UA and delete the prepopUA in PrePopulationDataAction. What do you think ?

Reply:
I'm not sure. If we are deleting it immediately, what is the point of having `request.prePopulationUserAnswers`? Cannot we just populate request.userAnswers directly?

Reply:
We agreed to copy over `request.prePopulationUserAnswers` to `request.userAnswers` and delete `request.prePopulationUserAnswers` after the cleanup (deletion of data that not needed in the new return).